File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int left = binarySearch(nums, target, true);
        int right = binarySearch( nums, target, false);
        return new int[]{left, right};
        }
    
    public int binarySearch(int[] nums, int target, boolean isLeftSearch){
        int start = 0;
        int end = nums.length-1;
        int result = -1;
        while(start <= end){
            int mid = start + (end-start)/2;
            if(nums[mid] < target){
                start = mid + 1;
            } else if(nums[mid] > target){
                end = mid - 1;
            } else {
                result = mid;
                if(isLeftSearch){
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }
        return result;
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def binarySearch(isLeftSearch):
            result = -1
            start = 0
            end = len(nums) - 1
            while start <= end:
                mid = start + (end-start)//2
                if nums[mid] < target:
                    start = mid + 1
                elif nums[mid] > target:
                    end = mid - 1
                else:
                    result = mid
                    if isLeftSearch:
                        end = mid - 1
                    else:
                        start = mid + 1
            return result
        return [binarySearch(True), binarySearch(False)]
                    
        